fluid/understand_sentiment_by_dynrnn.py

```diff
@@ -0,0 +1,193 @@
+
+try:
+    with open('word_dict.pkl', 'r') as f:
+        word_dict = cPickle.load(f)
+except:
+    word_dict = imdb.word_dict()
+    with open('word_dict.pkl', 'w') as f:
+        cPickle.dump(word_dict, f, cPickle.HIGHEST_PROTOCOL)
+
+
+def cache_reader(reader, clean):
+    print 'Reading data to memory'
+    fn = 'data.pkl'
+    if clean:
+        try:
+            os.remove(fn)
+        except:
+            pass
+    try:
+        with open(fn, 'r') as f:
+            items = cPickle.load(f)
+    except:
+        items = list(reader())
+        with open(fn, 'w') as f:
+            cPickle.dump(items, f, cPickle.HIGHEST_PROTOCOL)
+
+    print 'Done. data size %d' % len(items)
+
+    def __impl__():
+        offsets = range(len(items))
+        random.shuffle(offsets)
+        for i in offsets:
+            yield items[i]
+
+    return __impl__
+
+
+def crop_sentence(reader, crop_size):
+    unk_value = word_dict['<unk>']
+
+    def __impl__():
+        for item in reader():
+            if len([x for x in item[0] if x != unk_value]) < crop_size:
+                yield item
+
+    return __impl__
+
+
+def main():
+    args = parse_args()
+    data = fluid.layers.data(
+        name="words", shape=[1], lod_level=1, dtype='int64')
+    sentence = fluid.layers.embedding(
+        input=data, size=[len(word_dict), args.emb_dim])
+
+    sentence = fluid.layers.fc(input=sentence, size=200, act='tanh')
+
+    rnn = fluid.layers.DynamicRNN()
+    with rnn.block():
+        word = rnn.step_input(sentence)
+        lstm_size = 32
+        prev_hidden = rnn.memory(value=0.0, shape=[lstm_size])
+        prev_cell = rnn.memory(value=0.0, shape=[lstm_size])
+
+        def gate_common(
+                ipt,
+                hidden,
+                size, ):
+            gate0 = fluid.layers.fc(input=ipt, size=size, bias_attr=True)
+            gate1 = fluid.layers.fc(input=hidden, size=size, bias_attr=False)
+            gate = fluid.layers.sums(input=[gate0, gate1])
+            return gate
+
+        forget_gate = fluid.layers.sigmoid(
+            x=gate_common(word, prev_hidden, lstm_size))
+        input_gate = fluid.layers.sigmoid(
+            x=gate_common(word, prev_hidden, lstm_size))
+        output_gate = fluid.layers.sigmoid(
+            x=gate_common(word, prev_hidden, lstm_size))
+        cell_gate = fluid.layers.sigmoid(
+            x=gate_common(word, prev_hidden, lstm_size))
+
+        cell = fluid.layers.sums(input=[
+            fluid.layers.elementwise_mul(
+                x=forget_gate, y=prev_cell), fluid.layers.elementwise_mul(
+                    x=input_gate, y=cell_gate)
+        ])
+
+        hidden = fluid.layers.elementwise_mul(
+            x=output_gate, y=fluid.layers.tanh(x=cell))
+
+        rnn.update_memory(prev_cell, cell)
+        rnn.update_memory(prev_hidden, hidden)
+        rnn.output(hidden)
+
+    last = fluid.layers.sequence_pool(rnn(), 'last')
+    logit = fluid.layers.fc(input=last, size=2, act='softmax')
```

Comment:
We obey the convention that split model out of data pipeline, 
just like https://github.com/dzhwinter/benchmark/blob/master/fluid/resnet.py#L75 .